### PR TITLE
Partial refactor ShareableMixin into ShareableService

### LINKED ISSUE #11685: Sharing Pages returns 404
Steps to reproduce:
1) Users -> Pages -> Add new page -> Save
2) Click on `Share or Publish`
3) get 404 
image
 
That seems to work fine on 'org' and 'eu'. So either dev is broken or maybe I am doing smth wrong

EDIT:
can confirm, it works on 21.01, broken on dev

### PR DESCRIPTION
**Pull request by davelopez**
## What did you do? 
- Added a bunch of tests for testing the `sharing` API functionality with the relevant resource types (`Pages`, `Histories`, `Visualizations`). The tests are getting reused in each respective resource API by this [SharingApiTests](https://github.com/galaxyproject/galaxy/commit/87cbf8f5ecf6f61534883354f7f9c0193e7a963e#diff-5cbfa35916eed60f4ab5253024c3bd7a85ae8bb77626ec34c5fc3196737a8570R6) class.
- Created a new `ShareableService` that will replace the current `ShareableMixin` in favor of composition instead of inheritance. This is still a **work in progress** and right now is just a 'copy' (don't worry @jmchilton I promise it will be gone at some point 😆) of the Mixin with some minor modifications to make it work.
- Explicitly added the Pages `sharing` endpoint for [legacy](https://github.com/galaxyproject/galaxy/commit/fa09e9f0519bbdc13f18b82a4d1f9f707303bda9#diff-4a8e2455a0a6963c425766e46c399a63c23ecbbebc8d347514a08b3d56dbeabbR136-R159) and [FastAPI controllers](https://github.com/galaxyproject/galaxy/commit/fa09e9f0519bbdc13f18b82a4d1f9f707303bda9#diff-4a8e2455a0a6963c425766e46c399a63c23ecbbebc8d347514a08b3d56dbeabbR136-R159) as discussed in https://github.com/galaxyproject/galaxy/issues/11231 until we found a way of avoiding the duplication of the declaration of routes for the sharing functionality.
- This also should fix https://github.com/galaxyproject/galaxy/issues/11685 as now this `api/pages/{id}/sharing` endpoint is available again.

  image

### Additional comments
- The `Workflows` API probably should be adapted too, but I found out that it is overriding the sharing logic with a slightly different implementation and it is returning a `mako` template. I haven't looked in detail yet, but can I assume this is a leftover from previous refactorings and it should use the common sharing logic too or is this divergence intentional?

### TODO
- [x] Use the `ShareableService` to expose the `sharing` endpoint and remove `ShareableMixin` for:
  - [x] [Histories](https://github.com/galaxyproject/galaxy/pull/11712/commits/f42c979f94d136817212324a30944a2b7675b581#diff-a013b0d2d9763ac36ad6bb211fb06f0488adb2b1882390d5ca37835de8501fe1R584-R592)
  - [x] [Visualizations](https://github.com/galaxyproject/galaxy/pull/11712/commits/f58f5f489fdb3cb86775ec92a6ef8100b6189afc#diff-d8c5ee26f71bf3c7f8f22d55e53a622619e3f37127b84fde8a7edc76815f1f15R155-R164)
#### (Update) possibly out of scope of this PR
<del>- Add more tests for other sharing capabilities (like user-to-user sharing).</del>
<del>- Deal with the UI controllers using `ShareableMixin`. I am not familiar with them yet so I could use some advice on that :)</del>
<del>- Adapt the `Workflows` sharing logic to reuse the common `ShareableService` too?</del>
<del>- Completely remove `ShareableMixin` from the codebase.</del>

## Why did you make this change?
This is part of the effort to eliminate the old controller mixins, in this particular case `ShareableMixin`, described in https://github.com/galaxyproject/galaxy/issues/11231 that has been accelerated and incentivized by this bug https://github.com/galaxyproject/galaxy/issues/11685

## How to test the changes? 
(select the most appropriate option; if the latter, provide steps for testing below)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.

### CODE REVIEW

**Comment by davelopez** [comment]
At this point, *almost* all the **API controllers** that were using the `SharableMixin` (Pages, Visualizations, Histories) have been refactored to use the new `ShareableService` except for Workflows since it handles the `sharing` endpoint differently (still uses mako templates) and its Manager class will require further refactoring to implement `SharableModelManager` and a `SharableModelSerializer`.

On the other hand, the **UI controllers** will require more work to get rid of the `SharableMixin` since the rest of the methods in the mixin are more tied to their implementation and I am not sure yet how to proceed with them.

I think it may be a good idea to limit the scope of this PR to the current changes (with additional corrections or suggestions from any review) and aboard the rest of the changes related to the UI controllers in a separate PR. If this sounds like a good idea then It is ready for review :)

File: lib/galaxy/managers/sharable.py [hunk 1]
```diff
@@ -461,3 +471,222 @@ def _add_parsers(self):
             # chose by user should prob. only be available for admin? (most often we'll only need trans.user)
             # 'user'          : { 'op': ( 'eq' ), 'val': self.parse_id_list },
         })
+
+
+class SharingPayload(BaseModel):
+    action: str = Field(  # TODO: this seems like it should be a list of actions instead of separating them by '-'
```
Comment (mvdbeek) [inline lib/galaxy/managers/sharable.py:477]: This is not for this PR, but 💯 we should do that.

File: lib/galaxy/webapps/galaxy/api/histories.py [hunk 2]
```diff
@@ -577,3 +579,13 @@ def get_custom_builds_metadata(self, trans, id, payload=None, **kwd):
             'installed_builds': [{'label': ins, 'value': ins} for ins in installed_builds],
             'fasta_hdas': [{'label': f'{hda.hid}: {hda.name}', 'value': trans.security.encode_id(hda.id)} for hda in fasta_hdas],
         }
+
+    @expose_api
+    def sharing(self, trans, id, payload=None, **kwd):
+        """
+        * GET/POST /api/pages/{id}/sharing
+            View/modify sharing options for the page with the given id.
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/histories.py:587]: We should split this up into GET and POST and make the payload required for POST actions. But again, we can do that in another PR.

File: lib/galaxy/webapps/galaxy/api/page_revisions.py [hunk 3]
```diff
@@ -5,18 +5,13 @@
...
-    SharableItemSecurityMixin,
-    SharableMixin
-)
 from . import BaseGalaxyAPIController, depends
 
 log = logging.getLogger(__name__)
 
 
-class PageRevisionsController(BaseGalaxyAPIController, SharableItemSecurityMixin, UsesAnnotations, SharableMixin):
+class PageRevisionsController(BaseGalaxyAPIController):
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/page_revisions.py:14]: 🥇

**Review by mvdbeek — APPROVED** [review]
This is amazing stuff, thanks so much @davelopez!

**Comment by mvdbeek** [comment]
> **UI controllers**

I think we want to get rid of these. Is that something we can possibly work on in Q2 @dannon ?

File: lib/galaxy/webapps/galaxy/api/histories.py (hunk 2 above)
Comment (davelopez) [reply lib/galaxy/webapps/galaxy/api/histories.py:587]: Many thanks for the review!!

Yeah I agree, in the future FastAPI controller there will be two separate actions like it is already in Pages [here](https://github.com/galaxyproject/galaxy/pull/11712/files#diff-4a8e2455a0a6963c425766e46c399a63c23ecbbebc8d347514a08b3d56dbeabbR136-R160) but, since we will deprecate those legacy controllers at some point I tried to just add the bare minimum to the old ones 😅

**Comment by davelopez** [comment]
Thank you @mvdbeek!!
